### PR TITLE
Change ResolvedDependencyKey to a struct

### LINKED ISSUE #13138: ResolvedDependencyKey should be struct to reduce memory allocations
The `NuGet.Commands.ResolvedDependencyKey` class is allocated a lot during dependency resolution and instead could be a `struct` which would reduce memory usage.  This would be a breaking change for any user who is using the API.  This is a list of the public APIs that will change:

```
NuGet.Commands.ResolvedDependencyKey.Equals(NuGet.Commands.ResolvedDependencyKey other) -> bool
NuGet.Commands.ResolvedDependencyKey.ResolvedDependencyKey() -> void
static NuGet.Commands.ResolvedDependencyKey.operator !=(NuGet.Commands.ResolvedDependencyKey left, NuGet.Commands.ResolvedDependencyKey right) -> bool
static NuGet.Commands.ResolvedDependencyKey.operator ==(NuGet.Commands.ResolvedDependencyKey left, NuGet.Commands.ResolvedDependencyKey right) -> bool
```

Also, the `NuGet.Commands.UnexpectedDependencyMessages` was removed from the public API

### PR DESCRIPTION
**Pull request by Erarndt**
## Bug

Fixes: https://github.com/NuGet/Home/issues/13138

Regression? Last working version:

## Description

Switches `ResolvedDependencyKey` to a `struct` to avoid allocations just to do lookups in collections.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] Test exception - Existing tests provide coverage
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Review by jeffkl — DISMISSED** [review]

**Review by nkolev92 — CHANGES_REQUESTED** [review]
Changes actually look good, but I want to make sure we have a tracking issue created given that this is an SDK breaking change. 

By having an issue we can add a `Category;BreakingChange` label, so that the release notes tool can generate things like this: https://learn.microsoft.com/en-us/nuget/release-notes/nuget-6.8#nuget-sdk-breaking-changes

**Comment by jeffkl** [comment]
@nkolev92 I've filed https://github.com/NuGet/Home/issues/13138 and labeled it as a breaking change

@Erarndt I rebased your branch and pushed to resolve merge conflicts

File: src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,3 +1,8 @@
 #nullable enable
+NuGet.Commands.ResolvedDependencyKey.Equals(NuGet.Commands.ResolvedDependencyKey other) -> bool
+NuGet.Commands.ResolvedDependencyKey.ResolvedDependencyKey() -> void
+static NuGet.Commands.ResolvedDependencyKey.operator !=(NuGet.Commands.ResolvedDependencyKey left, NuGet.Commands.ResolvedDependencyKey right) -> bool
+static NuGet.Commands.ResolvedDependencyKey.operator ==(NuGet.Commands.ResolvedDependencyKey left, NuGet.Commands.ResolvedDependencyKey right) -> bool
+static NuGet.Commands.UnexpectedDependencyMessages.DependencyRangeHasMissingExactMatch(NuGet.Commands.ResolvedDependencyKey dependency) -> bool
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt]: Should we just make this method internal?

If we're making a breaking change, might as well break it completely. This doesn't need to be public.

I'm guessing this was the issue that @jeffkl fixed, so I'm looking for @NuGet/nuget-client thoughts.

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt]: `RestoreTargetGraph` exposes it at the moment: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs#L57

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt]: I was referring to the UnexpectedDependencyMessages method actually.


UnexpectedDependencyMessages does not need to be public at all. I was suggesting we change the method to internal.

**Comment by Erarndt** [comment]
> @nkolev92 I've filed [NuGet/Home#13138](https://github.com/NuGet/Home/issues/13138) and labeled it as a breaking change
> 
> @Erarndt I rebased your branch and pushed to resolve merge conflicts

Thanks for doing that

File: src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/PublicAPI/net472/PublicAPI.Unshipped.txt]: Ah sorry I thought you meant remove everything from public API.  I'll push a commit to remove this whole class from the public API and update the issue to indicate that the class was removed.

**Review by nkolev92 — APPROVED** [review]

**Review by jeffkl — APPROVED** [review]

**Pull request merged**